### DISCUSSION TITLE
Kura Build 5.4.0 RELEASE FAIL for profiles raspberry-pi-arm64 & raspberry-pi-arm64-nn with JAVA 17

### DISCUSSION THREAD

**Discussion by alexcorvis84** [eclipse-kura/kura#5113] [General]
Hi,

I'm trying to build the Kura 5.4.0 version from [latest tag](https://github.com/eclipse/kura/tree/KURA_5.4.0_RELEASE) for the profiles **raspberry-pi-arm64** & r**aspberry-pi-arm64-nn** with **JAVA 17** and I'm getting the following error which stops the compiling process:

![imagen](https://github.com/eclipse/kura/assets/44984326/0adac349-4d16-47fc-bf08-b8f79035cc1a)

I'm using the **./build-menu.sh** script available and selecting raspberry-pi-arm64, raspberry-pi-arm64-nn, Java 17 profiles. I'm performing the build process inside a Docker container with the image maven:3.8.5-openjdk-17.

![imagen](https://github.com/eclipse/kura/assets/44984326/47b4f39f-cbcc-474e-a6e4-7d7b4d4e7641)

Any help is welcome! 

Thanks in advance.

**Comment by marcellorinaldo** [comment]
> Hi,
> 
> I'm trying to build the Kura 5.4.0 version from [latest tag](https://github.com/eclipse/kura/tree/KURA_5.4.0_RELEASE) for the profiles **raspberry-pi-arm64** & r**aspberry-pi-arm64-nn** with **JAVA 17** and I'm getting the following error which stops the compiling process:
> 
> ![imagen](https://private-user-images.githubusercontent.com/44984326/301478220-0adac349-4d16-47fc-bf08-b8f79035cc1a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY3ODYxMzgsIm5iZiI6MTcwNjc4NTgzOCwicGF0aCI6Ii80NDk4NDMyNi8zMDE0NzgyMjAtMGFkYWMzNDktNGQxNi00N2ZjLWJmMDgtYjhmNzkwMzVjYzFhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjAxVDExMTAzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRhMzhiMzliMWFjNGQzOGM3N2E0YjYzYjY4ZDBmYWM4NjNmNDRmZGFkNjVjMzUzMDRlMmU4NzBhNTEwYWQwODQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.vgRzPQpcFH5xuZX-GW0DKpnoEj17t48J8p5RDg3Q0VU)
> 
> I'm using the **./build-menu.sh** script available and selecting raspberry-pi-arm64, raspberry-pi-arm64-nn, Java 17 profiles. I'm performing the build process inside a Docker container with the image maven:3.8.5-openjdk-17.
> 
> ![imagen](https://private-user-images.githubusercontent.com/44984326/301479451-47b4f39f-cbcc-474e-a6e4-7d7b4d4e7641.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY3ODYxMzgsIm5iZiI6MTcwNjc4NTgzOCwicGF0aCI6Ii80NDk4NDMyNi8zMDE0Nzk0NTEtNDdiNGYzOWYtY2JjYy00NzRlLWE2ZTQtN2Q3YjRkNGU3NjQxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjAxVDExMTAzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc3ZDNiMWJiYjVhODQ0MWM5MmUxNDdiNzU5OGE3YmI3ZjRiZTQwNGViZGFmOWZmZTg5ZWYwZGJjYTE1NzM2MTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.B9Nwl7Vpa05yvlDJtFdZVwRwjjIBiwWShUsY5nAAXjQ)
> 
> Any help is welcome!
> 
> Thanks in advance.

Hi @alexcorvis84. Currently to build Kura on a machine with JRE 17 you should use the maven profile `-Pjava17`, so that it will not compile bundles that are not compatbile with that Java version and adjust the test environment accordingly. I have not much experience with the `./build-menu.sh` script, but I think that it does not take the new flag into account.

Here some references to the relevant maven builds for building on Java 17 for target Java 17:

- [target platform](https://github.com/eclipse/kura/blob/d5c541b1c6ac55218b079d41af82063641d94aed/target-platform/pom.xml#L118)
- [kura main](https://github.com/eclipse/kura/blob/d5c541b1c6ac55218b079d41af82063641d94aed/kura/pom.xml#L244)
- [distrib](https://github.com/eclipse/kura/blob/d5c541b1c6ac55218b079d41af82063641d94aed/kura/distrib/pom.xml#L1155)

**Reply by alexcorvis84** [reply]
Thanks @marcellorinaldo for your quick reply. 

I know that profile -Pjava17 is required. Indeed, that's what I'm using selecting it through the ./build-menu.sh. It activates that profile, but it get stucked at the point I shared previously within the screenshot.

**Reply by alexcorvis84** [reply]
See below screenshot as evidence

![imagen](https://github.com/eclipse/kura/assets/44984326/41e73020-4f43-4470-80e2-9db504df7192)

**Reply by marcellorinaldo** [reply]
I see that it fails on the `features` build. In the image it fails when trying to get the dependencies for creating the script filter DP, which is skipped on a Java 17 build.

Unfourtunately, even excluding the script filter project it will fail on a later stage when producing the DP with the `osgi-dp` plugin. The `osgi-dp` plugin is compatible only with Java 8.

We discussed the update of it but the lack of expertise in Maven plugin development kept us postponing that issue.

This is the link to the `osgi-dp` plugin:
https://github.com/ctron/osgi-dp

**Reply by alexcorvis84** [reply]
Hi @marcellorinaldo, if I try to build **5.3.0 RELEASE** with the same method exposed, it works and I can get the binaries built, but with 5.4.0 RELEASE I got the error shared.

Is this the expected behaviour with 5.4.0 Release? I expected the same output as previous version... 🤔

**Reply by marcellorinaldo** [reply]
Inspecting closely, this commit might have broken the tests on java 17 by requiring the `wire.script.filter` dependency in the tests pom

- code point: https://github.com/eclipse/kura/blame/d26bca6c654a3285447b21eb979f26e28584244b/kura/test/org.eclipse.kura.rest.packages.provider.test/pom.xml#L427C15-L427C15
- git blame: https://github.com/eclipse/kura/commit/d74484bee2be0501f0d107ff01deec793e6ef384#diff-2ff5cbda7768db952d3525d7213b9350acee6699ec729ceb29595e5c5bfe3119

**Reply by mattdibi** [reply]
Ouff... my bad. I probably copy-pasted the integration tests from another bundle for starting out without checking. Obviously the integration tests for the `rest.packages` do not need the `wire.script.filter`. It didn't came out earlier because my toolchain (and the CI ones) are still Java8 based.

@alexcorvis84 We'll obviously need to fix this but as workaround I think that you can:
- Do not build the tests (`-Dmaven.test.skip=true`), I think this should work (hopefully)
- Switch to Java 8 for building.

**Comment by alexcorvis84** [comment]
> Ouff... my bad. I probably copy-pasted the integration tests from another bundle for starting out without checking. Obviously the integration tests for the `rest.packages` do not need the `wire.script.filter`. It didn't came out earlier because my toolchain (and the CI ones) are still Java8 based.
> 
> @alexcorvis84 We'll obviously need to fix this but as workaround I think that you can:
> 
>     * Do not build the tests (`-Dmaven.test.skip=true`), I think this should work (hopefully)
> 
>     * Switch to Java 8 for building.

Hi @mattdibi , build has worked with the two amends provided 👍🏼  Thanks for your suggestions! 

Anyway, I'll stay tuned for your news, hoping JAVA17 build gets fixed as soon as possible.

Many thanks in advance. Regards.

**Comment by mattdibi** [comment]
Fixed in #5117. @alexcorvis84 feel free to reopen this if you encounter issues with the fix.
